staging/src/k8s.io/apiserver/pkg/util/peerproxy/peer_discovery.go

```diff
@@ -0,0 +1,199 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package peerproxy
+
+import (
+	"fmt"
+	"net/http"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/client-go/discovery"
+	"k8s.io/klog/v2"
+
+	apidiscoveryv2 "k8s.io/api/apidiscovery/v2"
+	v1 "k8s.io/api/coordination/v1"
+	schema "k8s.io/apimachinery/pkg/runtime/schema"
+	apirequest "k8s.io/apiserver/pkg/endpoints/request"
+	responsewriterutil "k8s.io/apiserver/pkg/util/responsewriter"
+)
+
+func (h *peerProxyHandler) addPeerDiscoveryInfo(obj interface{}) {
+	serverIdentityLease, ok := obj.(*v1.Lease)
+	if !ok {
+		klog.ErrorS(fmt.Errorf("invalid lease object provided to addPeerDiscoveryInfo()"), "error adding peer served resources")
+		return
+	}
+
+	err := h.addPeerDiscoveryInfoToCache(serverIdentityLease)
+	if err != nil {
+		klog.ErrorS(err, "error adding peer served resources", "serverID", serverIdentityLease)
+	}
+}
+
+func (h *peerProxyHandler) deletePeerDiscoveryInfo(obj interface{}) {
+	h.peerAggDiscoveryCacheLock.Lock()
+	defer h.peerAggDiscoveryCacheLock.Unlock()
+	serverIdentityLease, ok := obj.(*v1.Lease)
+	if !ok {
+		klog.ErrorS(fmt.Errorf("invalid lease object provided to addPeerServedResources()"), "error deleting peer served resources")
+		return
+	}
+
+	h.peerAggDiscoveryCacheLock.Lock()
+	delete(h.peerAggDiscoveryResponseCache, serverIdentityLease.Name)
+	h.peerAggDiscoveryCacheLock.Unlock()
+}
+
+func (h *peerProxyHandler) updatePeerDiscoveryInfo(oldObj interface{}, newObj interface{}) {
+	oldLease, ok := oldObj.(*v1.Lease)
+	if !ok {
+		klog.Error("invalid lease object provided to updatePeerServedResources()")
+		return
+	}
+
+	newLease, ok := newObj.(*v1.Lease)
+	if !ok {
+		klog.Error("invalid lease object provided to updatePeerServedResources()")
+		return
+	}
+
+	switch {
+	case newLease == nil && oldLease == nil:
+		return
+	case newLease != nil && oldLease != nil:
+		// Delete old discovery info if holderIdentity changed, implying
+		// the sever was restarted.
+		if oldLease.Name == newLease.Name && *oldLease.Spec.HolderIdentity != *newLease.Spec.HolderIdentity {
+			h.peerAggDiscoveryCacheLock.Lock()
+			delete(h.peerAggDiscoveryResponseCache, oldLease.Name)
+			h.peerAggDiscoveryCacheLock.Unlock()
+		}
+
+		if err := h.addPeerDiscoveryInfoToCache(newLease); err != nil {
+			klog.ErrorS(err, "error adding peer discovery cache", "serverID", newLease.Name)
+		}
+	case oldLease == nil && newLease != nil:
+		if err := h.addPeerDiscoveryInfoToCache(newLease); err != nil {
+			klog.ErrorS(err, "error adding peer discovery cache", "serverID", newLease.Name)
+		}
+	case oldLease != nil && newLease == nil:
+		h.peerAggDiscoveryCacheLock.Lock()
+		delete(h.peerAggDiscoveryResponseCache, oldLease.Name)
+		h.peerAggDiscoveryCacheLock.Unlock()
```

Comment:
The lock/delete/unlock code appears in lots of places.  Would using a convention like `<functionName>Safe` and `<functionName>Unsafe` or similar to track which functions hold locks and which require the caller to hold the lock before calling be helpful?

Reply:
Moved the lock/unlock code to its own function `withPeerAggDiscoveryCacheLock()` - hopefully that helps reduce the noise and indicate that a lock will be acquired for the function